### ISSUE TITLE
kadalu consume a lot of CPU

### ISSUE THREAD

**Issue by vincentchong1** [kadalu/kadalu#660]
Will like to find out what is the reason kadalu is consuming a lot of CPU on the k8s worker node.

kadalu 0.8.4
Kubernetes v1.20.6
centos OS 3.10.0-1160.el7.x86_64

kubectl get pod -n kadalu -o wide
NAME                             READY   STATUS    RESTARTS   AGE   IP              NODE          NOMINATED NODE   READINESS GATES
kadalu-csi-nodeplugin-5wsmk      3/3     Running   0          39h   172.16.98.132   c432-kwo-02   <none>           <none>
kadalu-csi-nodeplugin-tnwld      3/3     Running   0          39h   172.16.162.69   c432-kwo-01   <none>           <none>
kadalu-csi-provisioner-0         5/5     Running   0          39h   172.16.98.131   c432-kwo-02   <none>           <none>
operator-56898988c5-tlcq6        1/1     Running   0          39h   172.16.162.65   c432-kwo-01   <none>           <none>
server-kadalu-storage-pool-0-0   1/1     Running   0          39h   172.16.162.68   c432-kwo-01   <none>           <none>
server-kadalu-storage-pool-1-0   1/1     Running   0          39h   172.16.98.130   c432-kwo-02   <none>           <none>

very high usage on this worker, kadalu/main.py is hover around 50%
[centos@C432-KWO-01 ~]$ ps -eo pid,ppid,cmd,%mem,%cpu --sort=-%cpu | head
  PID  PPID CMD                         %MEM %CPU
28618  6434 python3 /kadalu/main.py      0.1 49.7
 9315  8624 /opt/sbin/glusterfsd -N --v  0.0 19.9
13379     1 /usr/bin/kubelet --bootstra  0.6 14.8
13502  9290 /opt/sbin/glusterfs --proce  0.0 12.1
12894     1 /usr/bin/dockerd -H fd:// -  0.5  8.9
    1     0 /usr/lib/systemd/systemd --  0.0  7.0
20537     1 /usr/bin/perl -w /usr/bin/s  0.0  6.6
 4647  3076 /usr/local/bin/node /opt/tc  0.2  5.4
18358 18341 redis-server *:6379          0.0  4.9

but it is reasonable on this worker
[centos@C432-KWO-02 ~]$ ps -eo pid,ppid,cmd,%mem,%cpu --sort=-%cpu | head
  PID  PPID CMD                         %MEM %CPU
12753     1 /usr/bin/kubelet --bootstra  0.4 15.4
12273     1 /usr/bin/dockerd -H fd:// -  0.5  7.3
20339     1 /usr/bin/perl -w /usr/bin/s  0.0  6.8
 5942  5931 calico-node -felix           0.1  5.7
    1     0 /usr/lib/systemd/systemd --  0.0  4.3
 8504  8121 /opt/sbin/glusterfsd -N --v  0.1  3.8
 1617     1 /usr/bin/containerd          0.2  2.4
15692 14865 java -Dlogging.config=/opt/  1.0  1.8
18565 18504 java -Xmx1024m -Xms1024m -s  3.4  1.8

[centos@C432-KWO-01 ~]$ lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                4
On-line CPU(s) list:   0-3
Thread(s) per core:    1
Core(s) per socket:    1
Socket(s):             4
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel Xeon Processor (Skylake, IBRS)
Stepping:              4
CPU MHz:               2095.078
BogoMIPS:              4190.15
Hypervisor vendor:     KVM
Virtualization type:   full
L1d cache:             32K
L1i cache:             32K
L2 cache:              4096K
L3 cache:              16384K
NUMA node0 CPU(s):     0-3
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology eagerfpu pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 arat pku ospke avx512_vnni md_clear spec_ctrl intel_stibp

[centos@C432-KWO-02 ~]$ lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                4
On-line CPU(s) list:   0-3
Thread(s) per core:    1
Core(s) per socket:    1
Socket(s):             4
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel Xeon Processor (Skylake, IBRS)
Stepping:              4
CPU MHz:               2095.076
BogoMIPS:              4190.15
Hypervisor vendor:     KVM
Virtualization type:   full
L1d cache:             32K
L1i cache:             32K
L2 cache:              4096K
L3 cache:              16384K
NUMA node0 CPU(s):     0-3
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology eagerfpu pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 arat pku ospke avx512_vnni md_clear spec_ctrl intel_stibp

**Comment by leelavg** [comment]
@vincentchong1 thanks for raising the issue.

- First of it's kind which I'm aware wrt resources 
- Thanks for your first round of analysis, in any component `main.py` is the monitor and it's work is to restart subprocess if they are dying 
- Would you be able to trace out to which pod that specific process belongs to? Basically I can see three (operator, server, nodeplugin) kadalu pods running on that worker.
- It'd be helpful if you can login to kadalu specific container in those pods and pull out metrics from those.
- And wherever you see the high cpu, look for the logs for repeatedly terminating process


All in all, 50% is too much, will definitely look into it on priority, btw we are almost done with #643 which exposes metrics for prometheus and for other monitoring tools as a simple json file.

EDIT: My mistake, `start.py` is the monitor process and `main.py` corresponds to majority of the tasks for each components.

**Comment by aravindavk** [comment]
> @vincentchong1 thanks for raising the issue.
> 
> * First of it's kind which I'm aware wrt resources
> * Thanks for your first round of analysis, in any component `main.py` is the monitor and it's work is to restart subprocess if they are dying
> * Would you be able to trace out to which pod that specific process belongs to? Basically I can see three (operator, server, nodeplugin) kadalu pods running on that worker.

Based on `ps` output it is from Server.(`glusterfsd` process is there)

> * It'd be helpful if you can login to kadalu specific container in those pods and pull out metrics from those.
> * And wherever you see the high cpu, look for the logs for repeatedly terminating process
> 
> All in all, 50% is too much, will definitely look into it on priority, btw we are almost done with #643 which exposes metrics for prometheus and for other monitoring tools as a simple json file.

**Comment by leelavg** [comment]
- I doubt that, cause, the o/p is from worker node, not from the container and we can see `glusterfsd` on node-2 as well but no CPU hike from `main.py` on that.
- Either ways, unless we've logs/`ps` o/p from containers, we can't determine with certainity ig 😅

**Comment by aravindavk** [comment]
`ps` command shows CPU usage is currently expressed as the percentage of time spent running during the entire lifetime of a process.

Try using `top` command with `-n 2`. Example: `top -b -n 2 -d 0.2`

`top` command confirms if CPU usage is still high or it is reduced now. Even as a average, CPU usage looks high. We will look into that.

**Comment by vincentchong1** [comment]
my suspicious is the kadalu operator-56898988c5-tlcq6.

follow are data from operator container
```
[centos@C432-KMA-01 ~]$ kubectl exec -ti -n kadalu operator-56898988c5-tlcq6 -- bash
root@operator-56898988c5-tlcq6:/# top
top - 05:52:13 up 1 day, 23:08,  0 users,  load average: 10.54, 11.00, 9.87
Tasks:   4 total,   2 running,   2 sleeping,   0 stopped,   0 zombie
%Cpu(s): 38.3 us, 21.0 sy,  0.1 ni, 36.3 id,  0.3 wa,  0.0 hi,  3.7 si,  0.4 st
MiB Mem :  32011.9 total,  17677.0 free,  10977.0 used,   3358.0 buff/cache
MiB Swap:      0.0 total,      0.0 free,      0.0 used.  19954.3 avail Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
  203 root      20   0   66696  51704   6408 R  54.2   0.2 126:18.86 python3
    1 root      20   0   18236   6592   1236 S   0.0   0.0   0:14.98 python3
  233 root      20   0    4232   2176   1612 S   0.0   0.0   0:00.02 bash
  242 root      20   0    6080   1820   1320 R   0.0   0.0   0:00.00 top

root@operator-56898988c5-tlcq6:/# ps -ef | grep 203
root       203     1 44 01:07 ?        02:06:30 python3 /kadalu/main.py

root@operator-56898988c5-tlcq6:/# ps -ef | grep kadalu
root         1     0  0 Sep21 ?        00:00:14 python3 /kadalu/start.py
root       203     1 44 01:07 ?        02:06:35 python3 /kadalu/main.py

root@operator-56898988c5-tlcq6:/# top -b -n 2 -d 0.2
top - 05:54:15 up 1 day, 23:10,  0 users,  load average: 12.85, 11.87, 10.33
Tasks:   4 total,   2 running,   2 sleeping,   0 stopped,   0 zombie
%Cpu(s): 33.3 us, 18.5 sy,  0.0 ni, 44.4 id,  0.0 wa,  0.0 hi,  3.7 si,  0.0 st
MiB Mem :  32011.9 total,  17668.1 free,  10983.0 used,   3360.8 buff/cache
MiB Swap:      0.0 total,      0.0 free,      0.0 used.  19948.3 avail Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
  203 root      20   0   66696  51704   6408 R  53.3   0.2 127:20.69 python3
    1 root      20   0   18236   6592   1236 S   0.0   0.0   0:14.99 python3
  233 root      20   0    4232   2200   1628 S   0.0   0.0   0:00.02 bash
  249 root      20   0    5960   1724   1300 R   0.0   0.0   0:00.00 top

top - 05:54:15 up 1 day, 23:10,  0 users,  load average: 12.85, 11.87, 10.33
Tasks:   4 total,   2 running,   2 sleeping,   0 stopped,   0 zombie
%Cpu(s): 20.8 us,  6.5 sy,  0.0 ni, 68.8 id,  0.0 wa,  0.0 hi,  3.9 si,  0.0 st
MiB Mem :  32011.9 total,  17668.1 free,  10983.0 used,   3360.8 buff/cache
MiB Swap:      0.0 total,      0.0 free,      0.0 used.  19948.3 avail Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
  203 root      20   0   66696  51704   6408 R  45.0   0.2 127:20.78 python3
    1 root      20   0   18236   6592   1236 S   0.0   0.0   0:14.99 python3
  233 root      20   0    4232   2200   1628 S   0.0   0.0   0:00.02 bash
  249 root      20   0    5960   1724   1300 R   0.0   0.0   0:00.00 top
```

only 5 containers related to kadalu are running off kwo1 workers nodes
```
[centos@C432-KMA-01 ~]$ kubectl get pod -n kadalu -o wide | grep kwo-01
kadalu-csi-nodeplugin-tnwld      3/3     Running   0          46h   172.16.162.69   c432-kwo-01   <none>           <none>
operator-56898988c5-tlcq6        1/1     Running   0          46h   172.16.162.65   c432-kwo-01   <none>           <none>
server-kadalu-storage-pool-0-0   1/1     Running   0          46h   172.16.162.68   c432-kwo-01   <none>           <none>
```
Follow are information off kwo-01
```
[centos@C432-KWO-01 ~]$  ps -eo pid,ppid,cmd,%mem,%cpu --sort=-%cpu | head
  PID  PPID CMD                         %MEM %CPU
14305  6434 python3 /kadalu/main.py      0.1 44.2
 9315  8624 /opt/sbin/glusterfsd -N --v  0.0 17.7
13379     1 /usr/bin/kubelet --bootstra  0.6 14.9
13502  9290 /opt/sbin/glusterfs --proce  0.0 10.6
12894     1 /usr/bin/dockerd -H fd:// -  0.5  8.9
    1     0 /usr/lib/systemd/systemd --  0.0  7.1
20537     1 /usr/bin/perl -w /usr/bin/s  0.0  6.6
 4647  3076 /usr/local/bin/node /opt/tc  0.2  5.5
 1601     1 /usr/bin/containerd          0.2  3.8

[centos@C432-KWO-01 ~]$ top
top - 05:47:31 up 1 day, 23:03,  1 user,  load average: 12.64, 11.21, 9.44
Tasks: 1004 total,   4 running, 998 sleeping,   0 stopped,   2 zombie
%Cpu(s): 32.9 us, 24.7 sy,  2.4 ni, 36.5 id,  0.0 wa,  0.0 hi,  3.5 si,  0.0 st
KiB Mem : 32780208 total, 18095888 free, 11256160 used,  3428160 buff/cache
KiB Swap:        0 total,        0 free,        0 used. 20417472 avail Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
14305 root      20   0   66696  51704   6408 R  57.9  0.2 123:54.23 python3
    1 root      20   0  194424   6348   1548 S  15.8  0.0 201:01.87 systemd
12894 root      20   0 3016720 186652   6704 S  15.8  0.6 252:30.85 dockerd
20122 centos    20   0  163024   3136   1544 R  15.8  0.0   0:00.05 top
20137 root      20   0  903424  12740   5616 S  15.8  0.0   0:00.03 runc
13379 root      20   0 3212476 200852  21600 S  10.5  0.6 418:39.82 kubelet
 1601 root      20   0 1193104  67300   6696 S   5.3  0.2 107:31.50 containerd
 2182 root      20   0 1229688  55512   5756 S   5.3  0.2  22:32.55 node
 4647 root      20   0  345484  78220   6736 S   5.3  0.2 152:48.61 node
 7430 root      20   0  291380  30008   4720 S   5.3  0.1  19:33.88 node
 9641 root      20   0  640876  21676   3180 S   5.3  0.1   1:40.38 python3
19894 root      20   0  293572  34480   6036 S   5.3  0.1  20:10.61 node
20145 root      20   0   21408   7404   3148 S   5.3  0.0   0:00.01 runc:[0:PARENT]
20537 root      20   0  151848  11036   1052 S   5.3  0.0 187:56.38 sec
23175 root      20   0  713824  14192   3092 S   5.3  0.0   7:51.04 containerd-shim
26101 centos    20   0 2380832  20608   1280 S   5.3  0.1   8:48.07 rtmgr
    2 root      20   0       0      0      0 S   0.0  0.0   0:00.13 kthreadd
    4 root       0 -20       0      0      0 S   0.0  0.0   0:00.00 kworker/0:0H
    6 root      20   0       0      0      0 S   0.0  0.0   8:00.42 ksoftirqd/0
    7 root      rt   0       0      0      0 S   0.0  0.0   0:44.64 migration/0
    8 root      20   0       0      0      0 S   0.0  0.0   0:00.00 rcu_bh
    9 root      20   0       0      0      0 R   0.0  0.0  13:58.87 rcu_sched
   10 root       0 -20       0      0      0 S   0.0  0.0   0:00.00 lru-add-drain
   11 root      rt   0       0      0      0 S   0.0  0.0   0:06.68 watchdog/0

[centos@C432-KWO-01 ~]$ ps -ef | grep kadalu
root      6434  6414  0 Sep21 ?        00:00:14 python3 /kadalu/start.py
root      8624  8603  0 Sep21 ?        00:00:15 python3 /kadalu/server.py
root      9095  9071  0 Sep21 ?        00:00:04 /csi-node-driver-registrar --v=5 --csi-address=/plugin/csi.sock --kubelet-registration-path=/var/lib/kubelet/plugins_registry/kadalu/csi.sock
root      9290  9264  0 Sep21 ?        00:00:16 python3 /kadalu/start.py
root      9315  8624 17 Sep21 ?        08:14:21 /opt/sbin/glusterfsd -N --volfile-id kadalu-storage-pool.c432-kwo-01.bricks-kadalu-storage-pool-data-brick -p /var/run/gluster/glusterfsd-bricks-kadalu-storage-pool-data-brick.pid -S /var/run/gluster/brick.socket --brick-name /bricks/kadalu-storage-pool/data/brick -l - --xlator-option *-posix.glusterd-uuid=node-0 --process-name brick --brick-port 24007 --xlator-option kadalu-storage-pool-server.listen-port=24007 -f /kadalu/volfiles/kadalu-storage-pool.c432-kwo-01.bricks-kadalu-storage-pool-data-brick.vol
root      9641  9290  0 Sep21 ?        00:01:40 python3 /kadalu/main.py
root      9642  9290  0 Sep21 ?        00:00:24 python3 /kadalu/exporter.py
root      9643  9290  0 Sep21 ?        00:00:00 bash /kadalu/watch-vol-changes.sh
root     13502  9290 10 Sep21 ?        04:56:24 /opt/sbin/glusterfs --process-name fuse -l /var/log/gluster/gluster.log --volfile-id kadalu-storage-pool --fs-display-name kadalu:kadalu-storage-pool -f /kadalu/volfiles/kadalu-storage-pool.client.vol /mnt/kadalu-storage-pool
root     14305  6434 44 01:07 ?        02:05:10 python3 /kadalu/main.py

[centos@C432-KWO-01 ~]$ sudo docker ps | grep kadalu
45fcec6b9fe3   69593048aa3a           "/bin/sh -c 'while t…"   46 hours ago   Up 46 hours             k8s_kadalu-logging_kadalu-csi-nodeplugin-tnwld_kadalu_73bfb96b-5403-49cc-a23a-986f64fc7a1e_0
29d5c5a7380a   e9bb6eb56b6c           "python3 /kadalu/sta…"   46 hours ago   Up 46 hours             k8s_kadalu-nodeplugin_kadalu-csi-nodeplugin-tnwld_kadalu_73bfb96b-5403-49cc-a23a-986f64fc7a1e_0
d40e25daa61f   b34f79da08d5           "/csi-node-driver-re…"   46 hours ago   Up 46 hours             k8s_csi-node-driver-registrar_kadalu-csi-nodeplugin-tnwld_kadalu_73bfb96b-5403-49cc-a23a-986f64fc7a1e_0
77570746385e   k8s.gcr.io/pause:3.2   "/pause"                 46 hours ago   Up 46 hours             k8s_POD_kadalu-csi-nodeplugin-tnwld_kadalu_73bfb96b-5403-49cc-a23a-986f64fc7a1e_0
7eb22f774999   7009eca12eb9           "python3 /kadalu/ser…"   46 hours ago   Up 46 hours             k8s_server_server-kadalu-storage-pool-0-0_kadalu_23673f63-28d6-4631-bb76-14b2d7b5ade8_0
52e6f4e63da1   k8s.gcr.io/pause:3.2   "/pause"                 46 hours ago   Up 46 hours             k8s_POD_server-kadalu-storage-pool-0-0_kadalu_23673f63-28d6-4631-bb76-14b2d7b5ade8_0
d3fd0317e599   f9602e7fe3ef           "python3 /kadalu/sta…"   46 hours ago   Up 46 hours             k8s_kadalu-operator_operator-56898988c5-tlcq6_kadalu_9921e03a-0460-4f53-8168-e10c05d1ba6f_0
f63a1f094add   k8s.gcr.io/pause:3.2   "/pause"                 46 hours ago   Up 46 hours             k8s_POD_operator-56898988c5-tlcq6_kadalu_9921e03a-0460-4f53-8168-e10c05d1ba6f_0
```

leelavg: edited for readability.

**Comment by vincentchong1** [comment]
`[centos@C432-KWO-01 ~]$ pstree | grep -B2 -A3 python3

        |-containerd-shim-+-posagent-config---inotifywait
        |                 `-11*[{containerd-shim}]
        |-containerd-shim-+-python3---python3
        |                 `-11*[{containerd-shim}]
        |-containerd-shim-+-container_start---sleep
        |                 |-2*[morelater---5*[{morelater}]]
        |                 |-rtmRun---esmlc---98*[{esmlc}]
        |                 `-11*[{containerd-shim}]
        |-containerd-shim-+-python3---glusterfsd---31*[{glusterfsd}]
        |                 `-11*[{containerd-shim}]
        |-6*[containerd-shim-+-redis-server---4*[{redis-server}]]
        |                    `-11*[{containerd-shim}]]
        |-containerd-shim-+-csi-node-driver---10*[{csi-node-driver}]
        |                 `-11*[{containerd-shim}]
        |-containerd-shim-+-python3-+-bash---inotifywait
        |                 |         |-glusterfs---11*[{glusterfs}]
        |                 |         |-python3---8*[{python3}]
        |                 |         `-python3---{python3}
        |                 `-11*[{containerd-shim}]
        |-containerd-shim-+-tail
        |                 `-11*[{containerd-shim}]
`

**Comment by vincentchong1** [comment]
Is this context switch too high?

[root@C432-KWO-01 ~]# cat /proc/14305/status | grep ctxt
voluntary_ctxt_switches:        8146699
nonvoluntary_ctxt_switches:     1675677

**Comment by leelavg** [comment]
@vincentchong1 the stats are good enough and it states the problem is in `operator/main.py`

- Initial deductions:
1. Watch stream may be restarting every 30s https://github.com/kadalu/kadalu/pull/588 and that's making a good amount of context switches? https://github.com/kadalu/kadalu/blob/68ef132d7b2b85865aec291d676323c86b6bd5ba/operator/main.py#L872-L874
2. If the bottleneck is at https://github.com/kadalu/kadalu/blob/68ef132d7b2b85865aec291d676323c86b6bd5ba/operator/main.py#L846-L850 then it may be beyond Kadalu controller or needs an update to python k8s package?

- Basically nothing fancy happens in `operator/main.py`, subscribe to event stream and respond if there is an interesting event.
- Can you please provide operator logs in full length?

**Comment by vincentchong1** [comment]
log for
```
kadalu        operator-56898988c5-tlcq6                          1/1     Running     0          2d7h   172.16.162.65    c432-kwo-01   <none>           <none>
```
[kadalu-issues-660_operator-log.txt](https://github.com/kadalu/kadalu/files/7218617/kadalu-issues-660_operator-log.txt)

**Comment by leelavg** [comment]
Spot on! Some unknown (new) error in k8s watch is hammering CPU. Need to be fixed at the earliest. 

@vincentchong1 , pardon, due to timezone diff, pls expect a delayed response. From when are you facing this issue?

Basically is it the first time you're evaluating Kadalu or been using it from sometime?

@aravindavk if you've bandwidth can you take it up or else I'll look into it tomorrow as my network seems to be unreliable right now.

**Comment by vincentchong1** [comment]
> Basically is it the first time you're evaluating Kadalu or been using it from sometime?

we have used it since 0.8.3, recently upgraded to 0.8.4.  we use it for all of our pvc set up.

**Comment by leelavg** [comment]
- Can you please confirm whether you are behind a proxy or not
- If yes, after the referenced fix, you need to set `HTTP_PROXY` to URL of the proxy before deploying operator manifest like
`curl -Ls https://github.com/kadalu/kadalu/releases/latest/download/kadalu-operator.yaml | sed -e 's/"HTTP_PROXY"/$URL/' | kubectl apply -f -`
- Well, you get the idea wrt applying different manifests, all you need is to set `HTTP_PROXY` env value.

**Comment by vincentchong1** [comment]
yes, I think this is what you asked
```
$ kubectl get pod -n kube-system -o wide | grep proxy
haproxy-ingress-766d4bb9cf-6g6gw           1/1     Running   1          3d7h   172.16.162.111   c432-kwo-01   <none>           <none>
kube-proxy-c42nh                           1/1     Running   0          3d8h   172.28.19.36     c432-kma-01   <none>           <none>
kube-proxy-fhrxl                           1/1     Running   0          3d8h   172.28.19.39     c432-kwo-01   <none>           <none>
kube-proxy-t8rws                           1/1     Running   0          3d8h   172.28.19.40     c432-kwo-02   <none>           <none>
```

**Comment by leelavg** [comment]
- No, these are part of control plane
- The proxy I meant is, for accessing api server itself
- Do you see any proxy setup in kubeconfig file? If yes, then that should be used, if no, issue need to be revisited 😔

**Comment by vincentchong1** [comment]
let me know how I can check it out.
our underlying cluster networking is based on calico, and follow are the port 10250 info
```
[centos@C432-KMA-01 ~]$ sudo netstat -atupn | grep -E "10250|6443" | grep -i listen
tcp6       0      0 :::10250                :::*                    LISTEN      12687/kubelet
tcp6       0      0 :::6443                 :::*                    LISTEN      18006/kube-apiserve
[centos@C432-KMA-01 ~]$ ssh kwo1 'sudo netstat -atupn | grep -E "10250|6443"'
tcp        0      0 172.28.19.39:20360      172.28.19.36:6443       ESTABLISHED 14466/kube-proxy
tcp        0      0 172.28.19.39:18722      172.28.19.36:6443       ESTABLISHED 13379/kubelet
tcp6       0      0 :::10250                :::*                    LISTEN      13379/kubelet
[centos@C432-KMA-01 ~]$ ssh kwo2 'sudo netstat -atupn | grep -E "10250|6443"'
tcp        0      0 172.28.19.40:65466      172.28.19.36:6443       ESTABLISHED 12753/kubelet
tcp        0      0 172.28.19.40:64436      172.16.98.148:8993      TIME_WAIT   -
tcp        0      0 172.28.19.40:10768      172.28.19.36:6443       ESTABLISHED 13697/kube-proxy
tcp6       0      0 :::10250                :::*                    LISTEN      12753/kubelet

```

**Comment by leelavg** [comment]
- Sry, not this o/p. Can you've a look at https://github.com/kubernetes-client/python/issues/549
- what does `grep -i proxy ~/.kube/config` say, pls remove any certs/creds from o/p?

**Comment by leelavg** [comment]
- btw, if you know a bit of python, as a workaround 
- `k exec -it deploy/operator -n kadalu -- sed -e -i 's/ProtocolError/(ProtocolError, NewConnectionError)/g' /kadalu/main.py`
- `k exec -it deploy/operator-n kadalu -- kill -9 <pid of main.py>`

**Comment by Expro** [comment]
I'm seeing same thing - Kadulu, even when idle (not provisioning volumes) consumes a lot of CPU on 0.8.4: around 50 - 70% of core.

```
ps -eo pid,ppid,cmd,%mem,%cpu --sort=-%cpu | head
    PID    PPID CMD                         %MEM %CPU
   3594    3492 python3 /kadalu/main.py      0.3 61.6
3302129       1 /usr/lib/systemd/systemd-ho  0.0  7.0
3302114       1 /usr/lib/systemd/systemd --  0.0  3.2
   1332       1 /usr/bin/kubelet --bootstra  0.7  2.4
   1008     760 /usr/libexec/netdata/plugin  0.1  0.9
    760       1 /usr/sbin/netdata -P /var/r  0.9  0.8
  10026   10014 /jellyfin/jellyfin           2.7  0.8
   8996    4767 /opt/sbin/glusterfs --proce  0.0  0.6
   2441    1768 ./kube-ovn-daemon --ovs-soc  0.4  0.5
```

I'm not behind proxy.
